haskell/defs.bzl

```
@@ -53,6 +57,7 @@ _haskell_common_attrs = {
     "deps": attr.label_list(
         aspects = [haskell_cc_libraries_aspect],
     ),
+    "narrowed_deps": attr.label_list(providers = [HaskellLibraryInfo]),
```

```suggestion
    "narrowed_deps": attr.label_list(providers = [HaskellInfo, HaskellLibraryInfo]),
```
IIUC this needs to provide `HaskellInfo` for the modules information, and `HaskellLibraryInfo`, because it should be a `haskell_library`, correct?

---

That is correct. The reason I didn't add `HaskellInfo` is that I expect it to be present whenever `HaskellLibraryInfo` is.

Also, isn't the quality of the error message better with a fail call? Is the user supposed to know which rules provide which providers?

---

> Also, isn't the quality of the error message better with a fail call? Is the user supposed to know which rules provide which providers?

Fair enough, that is an issue. The error message takes the form
> in deps attribute of haskell_library rule //some:target: '//some:dep' does not have mandatory providers: ['HaskellInfo', 'HaskellLibraryInfo']

which is indeed not terribly informative. Happy to go with a `fail` approach instead.

---

I tried `fail` in 1095d23.